src/main/java/org/ice4j/stack/NetAccessManager.java

```diff
@@ -47,6 +48,24 @@
     private static final Logger logger
         = Logger.getLogger(NetAccessManager.class.getName());
 
+    /**
+     * Thread pool to execute {@link MessageProcessor}s across all
+     * {@link NetAccessManager}s.
+     * The work-stealing thread pool tries to utilizes all available processors
+     * in parallel, but does not have guarantee about the order of execution,
+     * but underlying transport (UDP) also does not have such guarantee,
+     * so it is ok when some of the messages will be processed out of
+     * arrival (enqueuing) order, but faster.
+     */
```

Comment:
@bbaldino yes, I do agree that this need to be considered. There is an option to use different executor service (for example `Executors.newFixedThreadPool`), which has one queue for all processing threads, instead of `ExecutorService` which has different queues per each thread. But I think conceptually even when queue is single, imagine if you have queued `pkt1` and `pkt2` into queue: `thread 1` takes `pkt1` from queue and then `thread 2` immediately takes `pkt2` from queue, then for some reason OS kernel decides to suspend `thread 1` execution, but continue with execution of `thread 2`, so it is likely `pkt2` will be handled before `pkt1` even so it was popped form queue first... So, the only way to have truly sequential in order execution is to have single thread, but earlier there was already created `3` threads to handle single queue, it was already possible that items were put into the queue in one order, but item processing is effectively done in different order, because earlier there was a lot more threads which OS kernel were switching a lot more aggressively.

Reply:
Understood, but did we have multiple threads reading from this queue before?  I thought it was only a single thread (MessageProcessor had a queue which it read from a single thread, no?)

Reply:
> Understood, but did we have multiple threads reading from this queue before?

@bbaldino Yes, the queue was single per connected peer. Actually queue was [created per `NetAccessManager`](https://github.com/jitsi/ice4j/blob/513fb588d65de5a1437f50c51313a98a1fe18e58/src/main/java/org/ice4j/stack/NetAccessManager.java#L80), which is created [per `StunStack`](https://github.com/jitsi/ice4j/blob/513fb588d65de5a1437f50c51313a98a1fe18e58/src/main/java/org/ice4j/stack/StunStack.java#L407), `StunStack` is created [per `Agent`](https://github.com/jitsi/ice4j/blob/513fb588d65de5a1437f50c51313a98a1fe18e58/src/main/java/org/ice4j/ice/Agent.java#L1338) and `Agent` is created for each peer connected to `JVB`.
By default there was [`3` MessageProcessors](https://github.com/jitsi/ice4j/blob/513fb588d65de5a1437f50c51313a98a1fe18e58/src/main/java/org/ice4j/stack/StunStack.java#L46) (which was basically custom Thread implementation) [created](https://github.com/jitsi/ice4j/blob/513fb588d65de5a1437f50c51313a98a1fe18e58/src/main/java/org/ice4j/stack/NetAccessManager.java#L512) to [process `RawMessage`s from single queue](https://github.com/jitsi/ice4j/blob/513fb588d65de5a1437f50c51313a98a1fe18e58/src/main/java/org/ice4j/stack/NetAccessManager.java#L522).
This results in extremely large number of threads created to process STUN messages. For `30` audio-only conferences with `3` peers each there was created `270` threads to process stun messages, no matter that physical machine is only have for example `4` cores and there is no reason to create more threads than number of cores for computational work like decoding stun messages.

>  I thought it was only a single thread (MessageProcessor had a queue which it read from a single thread, no?)

Actually no, even so that it was possible to set [default pool size to `1`](https://github.com/jitsi/ice4j/blob/513fb588d65de5a1437f50c51313a98a1fe18e58/src/main/java/org/ice4j/stack/StunStack.java#L46) by replacing `DEFAULT_THREAD_POOL_SIZE` constant, it would anyway create `1` thread per connected peer. So for `30` conferences with `3` peers each it will result in `90` threads, on for example `4` cores machine.

The thread allocation strategy, when thread count is linear to number of connected peers does not scale well.
The number of CPU cores is fixed and when there is too many threads created, it results in extra pressure on OS kernel scheduler as well as on JVM itself, which slows down performance of system in general.
In all my PRs I'm reducing number of threads by putting more work on existing threads and by avoid blocking & waiting inside tasks executed on thread.

Reply:
where did we end up on this? do we still have the multiple threads?  a couple places you mentioned are per X, but we only ever have one instance of X per connection (one peer connection and everything is mux'd).

Reply:
> do we still have the multiple threads? 

Yes.
With moving from `ForkJoinPool` to regular pool there is again single queue of tasks to execute, but multiple (fixed count) executor threads.
Processing threads are managed by pool and it's number is equal to CPU cores.

> a couple places you mentioned are per X, but we only ever have one instance of X per connection (one peer connection and everything is mux'd).

The problem with old code was the number of processing threads were linear to number of peers connected to JVB, actually it was 3 thread for each peer. 
Number of processing threads is now fixed and equal to the number of CPU cores.